File: packages/calendar/src/gavel_calendar/agenda.py
"""Build the contract agenda (docs/CONTRACT.md §1) from a `ParsedInvite`.

Pure and offline: no network, no store, no ears. `session_id` and the
attendee email→discordId map are supplied by the caller because neither
comes from the .ics itself.
"""

from __future__ import annotations

from .ics_parser import ParsedInvite, TopicDraft
from .schema import EARS_DEFAULT_POLICY, ContractAgenda


def _purpose(invite: ParsedInvite, description: str) -> str:
    for line in description.splitlines():
        stripped = line.strip()
        if not stripped:
            continue
        # First bulleted/numbered line means the description moved into topics.
        if stripped[0] in "-*•" or (
            stripped[0].isdigit() and ("." in stripped[:3] or ")" in stripped[:3])
        ):
            break
        return stripped
    return invite.title


def _attendee_id(email: str, attendee_map: dict[str, str]) -> str:
    # Unmapped: the email is a stable, unique fallback id — it just will not
    # match a real Discord speaker until an operator maps it. See settings.py.
    return attendee_map.get(email.lower(), email)


def _budget_seconds(topics: list[TopicDraft], total_seconds: int) -> list[int]:
    """Explicit durations win; topics with none split what is left evenly."""
    known = [t.budget_seconds for t in topics if t.budget_seconds is not None]
    remaining = max(total_seconds - sum(known), 0)
    unknown_count = sum(1 for t in topics if t.budget_seconds is None)
    if unknown_count == 0:
        return [t.budget_seconds for t in topics]  # type: ignore[misc]

    share = remaining // unknown_count
    out: list[int] = []
    given = 0
    for t in topics:
        if t.budget_seconds is not None:
            out.append(t.budget_seconds)
            continue
        given += 1
        # last one absorbs the remainder so the split fully uses `remaining`
        out.append(remaining - share * (unknown_count - 1) if given == unknown_count else share)
    return out


def build_agenda(
    invite: ParsedInvite,
    session_id: str,
    attendee_map: dict[str, str] | None = None,
    policy_overrides: dict[str, float | bool] | None = None,
) -> dict:
    attendee_map = attendee_map or {}
    total_seconds = invite.duration_seconds

    attendees = [
        {
            "discordId": _attendee_id(a.email, attendee_map),
            "name": a.name,
            "role": "host" if a.is_organizer else "attendee",
        }
        for a in invite.attendees
    ]
    by_name = {
        a.name.strip().lower(): att["discordId"]
        for a, att in zip(invite.attendees, attendees, strict=True)
    }

    # Whoever called the meeting is the standing best guess for anything the
    # brief did not say. An agenda with no owner and nobody who must be heard
    # gives the chair nothing to chase — she can't ask "who owns this?" or
    # notice that the one person who had to speak never did. A guessed host is
    # wrong far less often than empty is useless, and the compose form shows
    # the guess so it can be corrected before the invite goes out.
    host_id = next(
        (a["discordId"] for a in attendees if a["role"] == "host"),
        attendees[0]["discordId"] if attendees else None,
    )

    budgets = _budget_seconds(invite.topics, total_seconds)
    topics = []
    for i, (draft, budget) in enumerate(zip(invite.topics, budgets, strict=True), start=1):
        owner_id = by_name.get(draft.owner_name.strip().lower()) if draft.owner_name else None
        if owner_id is None:
            owner_id = host_id
        # Unmatched names are dropped rather than passed through — the brain
        # resolves mustHear against discordIds, not free text.
        must_hear_ids = [
            by_name[name.strip().lower()]
            for name in draft.must_hear_names
            if name.strip().lower() in by_name
        ]
        if not must_hear_ids and owner_id is not None:
            # The owner of a topic is the one person who has to be heard on it.
            must_hear_ids = [owner_id]
        topics.append(
            {
                "id": f"t{i}",
                "title": draft.title,
                "goal": draft.goal,
                "budgetSeconds": budget,
                "owner": owner_id,
                "mustHear": must_hear_ids,
                "questions": list(draft.questions),
            }
        )

    agenda = {
        "sessionId": session_id,
        "purpose": _purpose(invite, invite.description),
        "totalSeconds": total_seconds,
        "attendees": attendees,
        "topics": topics,
    }
    # Only present when the invite genuinely overrides something — see
    # schema.py. `ears`'s Agenda.policy does not deep-merge a supplied dict,
    # so we merge the override onto its full default table ourselves: the
    # wire payload always carries all ten keys, only the overridden ones
    # changed, never a partial dict that would drop the rest for this session.
    if policy_overrides:
        agenda["policy"] = {**EARS_DEFAULT_POLICY, **policy_overrides}
    # Fails loudly, offline, before this ever reaches ears or the brain.
    ContractAgenda.model_validate(agenda)
    return agenda


def attendee_name(agenda: dict, discord_id: str | None) -> str:
    """The inverse of the `by_name` lookup above: a built agenda's `discordId`
    back to its display name, for anywhere an id must read as a human (a join
    page, an .ics `DESCRIPTION`). An unmapped id — the email fallback from
    `_attendee_id` — is returned as-is rather than blanked out.
    """
    if discord_id is None:
        return ""
    for a in agenda["attendees"]:
        if a["discordId"] == discord_id:
            return str(a["name"])
    return discord_id

File: packages/extension/src/lib/agenda.ts
// The confirm table → the contract agenda (docs/CONTRACT.md §1).
//
// A port of three pieces of `packages/calendar`, kept in the same order so a
// diff against the Python reads line for line:
//   compose.py:_fill_topic_minutes / _topic_drafts   → fillTopicMinutes / topicDrafts
//   agenda.py:build_agenda                           → buildAgenda
//   schema.py:EARS_DEFAULT_POLICY / ENFORCEMENT_LEVELS → the two tables below
//
// Pure: no DOM, no network, no chrome.*. `test/agenda.test.ts` pins it against
// the fixture values the Python tests use.

import { cleanMinutes } from "./sanitize.js";
import type {
  ConfirmState,
  ContractAgenda,
  ContractAttendee,
  ContractTopic,
  Enforcement,
  Person,
  Policy,
  TopicRow,
  TopicType,
} from "./types.js";

/** Mirrors `ears-discord`'s `DEFAULT_POLICY` via `schema.EARS_DEFAULT_POLICY`.
 *  The wire payload always carries every key: `ears`'s own `Agenda.policy` does
 *  not deep-merge, so a partial dict would silently drop the rest. */
export const EARS_DEFAULT_POLICY: Policy = {
  floorShareThreshold: 0.6,
  floorWindowSeconds: 120,
  softHandoverSeconds: 45,
  hardHandoverSeconds: 90,
  topicOverrunFactor: 1.2,
  silenceSeconds: 15,
  minSecondsBetweenInterventions: 45,
  offAgendaGraceSeconds: 20,
  allowMute: false,
  escalateAfterSeconds: 10,
  muteSeconds: 15,
  requireStart: false,
  timed: true,
};

/** The gauge on the confirm view — `schema.ENFORCEMENT_LEVELS`, verbatim. */
export const ENFORCEMENT_LEVELS: Record<Enforcement, Policy> = {
  low: {
    offAgendaGraceSeconds: 45,
    minSecondsBetweenInterventions: 90,
    topicOverrunFactor: 1.5,
    softHandoverSeconds: 90,
    hardHandoverSeconds: 0,
    silenceSeconds: 25,
    escalateAfterSeconds: 20,
    allowMute: false,
  },
  medium: {
    offAgendaGraceSeconds: 8,
    minSecondsBetweenInterventions: 45,
    topicOverrunFactor: 1.2,
    softHandoverSeconds: 45,
    hardHandoverSeconds: 90,
    silenceSeconds: 15,
    escalateAfterSeconds: 10,
    allowMute: false,
  },
  high: {
    offAgendaGraceSeconds: 5,
    minSecondsBetweenInterventions: 25,
    topicOverrunFactor: 1.05,
    softHandoverSeconds: 30,
    hardHandoverSeconds: 30,
    silenceSeconds: 10,
    escalateAfterSeconds: 6,
    allowMute: true,
    muteSeconds: 15,
  },
};

export const DEFAULT_ENFORCEMENT: Enforcement = "medium";

export function policyFor(level: Enforcement): Policy {
  return { ...EARS_DEFAULT_POLICY, ...(ENFORCEMENT_LEVELS[level] ?? ENFORCEMENT_LEVELS[DEFAULT_ENFORCEMENT]) };
}

/** `compose.py:_fill_topic_minutes`. Named durations are kept; the rest of the
 *  meeting splits evenly, in whole minutes, across the rows that did not say,
 *  and the last of those absorbs the remainder. Whole minutes matter: the
 *  description round-trips through `ics_parser`, which reads minutes only. */
export function fillTopicMinutes(minutes: (number | null)[], totalMinutes: number): number[] {
  const known = minutes.filter((m): m is number => m !== null);
  const remaining = Math.max(totalMinutes - known.reduce((a, b) => a + b, 0), 0);
  const unknownCount = minutes.length - known.length;
  if (unknownCount === 0) return known;
  const share = Math.floor(remaining / unknownCount);
  const extra = remaining - share * unknownCount;
  let given = 0;
  return minutes.map((m) => {
    if (m !== null) return m;
    given += 1;
    return share + (given === unknownCount ? extra : 0);
  });
}

export interface TopicDraft {
  title: string;
  budgetSeconds: number;
  ownerName: string | null;
  mustHearNames: string[];
  type: TopicType;
}

/** `compose.py:_topic_drafts` — blank-title rows are dropped, everything else
 *  gets an explicit whole-minute budget before it reaches `buildAgenda`. */
export function topicDrafts(rows: TopicRow[], totalMinutes: number): TopicDraft[] {
  const named = rows.filter((r) => r.title.trim());
  const minutes = named.map((r) => cleanMinutes(r.minutes));
  const filled = fillTopicMinutes(minutes, totalMinutes);
  return named.map((r, i) => ({
    title: r.title.trim(),
    budgetSeconds: filled[i]! * 60,
    ownerName: r.owner.trim() || null,
    mustHearNames: r.mustHear
      .split(",")
      .map((n) => n.trim())
      .filter(Boolean),
    type: r.type === "presentation" ? "presentation" : "discussion",
  }));
}

/** Twelve hex characters, like `uuid.uuid4().hex[:12]` in `compose.handle_send`. */
export function newSessionId(random: () => string = () => crypto.randomUUID()): string {
  return random().replace(/-/g, "").slice(0, 12);
}

export interface BuildInput {
  state: ConfirmState;
  /** The signed-in user — `role: "host"`. Added to the attendees if the editor
   *  did not list them (Google shows the organiser as a chip only sometimes). */
  host: Person;
  sessionId: string;
}

/** `agenda.py:build_agenda`, for one meeting. `discordId` is the e-mail — the
 *  same stable fallback the Python uses until `CALENDAR_ATTENDEE_MAP` maps it. */
export function buildAgenda({ state, host, sessionId }: BuildInput): ContractAgenda {
  const people: Person[] = state.attendees.some((a) => a.email === host.email)
    ? state.attendees
    : [host, ...state.attendees];

  const attendees: ContractAttendee[] = people.map((p) => ({
    discordId: p.email,
    name: p.name,
    role: p.email === host.email ? "host" : "attendee",
  }));

  const byName = new Map<string, string>();
  for (const a of attendees) byName.set(a.name.trim().toLowerCase(), a.discordId);
  const byEmail = new Map(attendees.map((a) => [a.discordId, a.discordId] as const));
  const resolve = (name: string): string | undefined => {
    const key = name.trim().toLowerCase();
    return byName.get(key) ?? byEmail.get(key) ?? uniqueFirstName(attendees, key);
  };

  const hostId = attendees.find((a) => a.role === "host")?.discordId ?? attendees[0]?.discordId ?? null;

  const totalSeconds = state.durationMinutes * 60;
  const topics: ContractTopic[] = topicDrafts(state.rows, state.durationMinutes).map((draft, i) => {
    const ownerId = (draft.ownerName ? resolve(draft.ownerName) : undefined) ?? hostId;
    // Unmatched names are dropped, not passed through — the brain resolves
    // mustHear against ids, not free text.
    let mustHear = draft.mustHearNames.map(resolve).filter((id): id is string => id !== undefined);
    if (mustHear.length === 0 && ownerId !== null) mustHear = [ownerId];
    return {
      id: `t${i + 1}`,
      title: draft.title,
      goal: "",
      budgetSeconds: draft.budgetSeconds,
      owner: ownerId,
      mustHear,
      questions: [],
      type: draft.type,
    };
  });

  // The host's own one-liner wins; the title stands in when there is none,
  // as `agenda._purpose` falls back to `invite.title`.
  const purpose = state.purpose.trim() || state.title.trim();
  return {
    sessionId,
    purpose,
    totalSeconds,
    attendees,
    topics,
    policy: policyFor(state.enforcement),
  };
}

/** The Python's `by_name` is exact. Chip labels off Google are full names while
 *  the model writes what the brief said ("Artem"), so a first name is accepted
 *  when it names exactly one person on the invite. */
function uniqueFirstName(attendees: ContractAttendee[], key: string): string | undefined {
  if (!key || key.includes(" ")) return undefined;
  const hits = attendees.filter((a) => a.name.trim().toLowerCase().split(/\s+/)[0] === key);
  return hits.length === 1 ? hits[0]!.discordId : undefined;
}

/** What `POST /ext/v1/agendas` refuses with a 422, checked here first so the
 *  panel never gets as far as the network with an empty agenda. */
export function isRegistrable(agenda: ContractAgenda): boolean {
  return agenda.topics.length > 0 && agenda.totalSeconds > 0 && agenda.attendees.length > 0;
}
